_posts/2018-09-07-Golang-with-The-Paal-Gang.markdown

```diff
@@ -0,0 +1,248 @@
+  * myArray :=  [<optional len>]<datatype>{<optional initializer>}
+  3. Array utils: 
+    -> len(myArray) => int length
+    -> for idx, val := range myArray {} // foreach with enumerate
+
+### Slices
+  1. Slices are parts of an array
+  2. array := [<len>]<datatype>{<optional initializer>}
+      * slice := array[<begin>:<end>]
+      * slices point to the underlying array
+  3. Slice utils:
+    -> len(slice)  // length of slice
+    -> cap(slice) // capacity of slice
+    -> append(<slice>, <value>[, <more values>]) // This returns a new instance
+      * When using append, it is adviced to return to the same slice as appended to
+slice = append(slice, value)
+
+### Maps
+  1. Maps are like python dictionaries, but they are type sensitive
+  2. myMap := map[<key datatype>]<value datatype>{<optional key>: <optional value>}
+    * forExample := map[string]float64{"myKey": 33}
+    * Getting/setting : myMap["myKey"] = <myValue>
+  3. Map utils
+     -> for <key>, <value> := range myMap {...}
+     -> for <key> := range myMap {...}
+
+## Working with datatypes
+
+Here comes some of the nice golang features into play.  So far we've only
+looked at the basic capabilities and mechanics of the language
+
+### Custom types
+  * Custom types is like typedef in C
+  * Syntax: type <MyType> <ActualType>
+  * Example: type Hours int
+  * Custom types can be compared to same custom type or underlying type, but cannot be compared with different custom even if they share the same underlying type
+
+### Methods
+
+But Go isn't object oriented?
+
+  * Methods are implicit methods on types.  It is a function with a specific receiver argument, which tells which types implement the method
+  * func (t <MyType>) <MyMethod>(<MyArgs>) <MyReturn> {...}
+  * Now <MyType> vars will implement the <MyMethod> method
+  * In Go it is common to write methods that gracefully handle being called with nil receiver
+  * Example: 
+```
+func (v Vertex) Abs() float64 {...}
+/* (using) --> */ vert.Abs()
+```
+
+### Structs
+
+
+* In concept pretty much like structs in C -> structured data
+* Convention to make all struct methods take either values or pointers as recv
+* Declaring struct:
+```
+      type MyStruct struct {
+         myArg string
+     }
+```
+* Initializing struct:
+```
+      myStruct := MyStruct{MyArg: "Hello World!"}
+      myStruct := MyStruct{"Hello World"} // Does the same, but need order of args
+```
+* Using struct:
+      -> Non cap members of struct are private
+      -> Accessing struct members are the same for pointers and value based structs
+          myStruct.MyArg
+
+### Interfaces
+
+  * The interface type is a set of method signatures
+  * Interfaces are implemented implicitly, if a datatype has the methods to satisfy the interface, it automatically implements the interface
+  * Syntax for declaring interface: type <MyInterface> interface {/*methods*/}
+  * Describing interfaces:  fmt.Printf("(%v, %T)\n", i, i) // i is an interface
```

Comment:
This describing interfaces text was just to show that interfaces are just and underlying datastructure with data on value and datatype of the *whatever* implementing the interface.  Omitting a long text about fmt here, but maybe a new chapter?

Reply:
Changed to

```
fmt.Printf("(%v, %T)\n", i, i) // i is an interface
```

This will show us that the interface is just a tuple of a value and datatype.
This is well illustrated in the [Tour of Go](https://tour.golang.org/methods/14) example.

```
Some interfaces:
(<nil>, <nil>)
(42, int)
(hello, string)
```